File: docs/palindromes-in-peg/generate_online_peg.py
"""Generate the derived FIFO/whole-round PAL candidate on original input.

No input expansion, recognition budget, or word-length cap is accepted.
The compilation memory limit is a host resource limit only. Output is renamed
into place only after every ordinary PEG rule has been emitted. Source
correctness remains subject to the contracts in TRANSLATION_STRATEGY.md.
"""
import argparse
import gc
import hashlib
import json
from pathlib import Path
import resource
import signal
import time

from galil_clock import derive
from scaffold_circuit_galil import build_online
from scaffold_event_buffer import buffer_source, pack_service
from symbolic_sca2peg import share_expressions
import scaffold_artifact


def source_signature():
  root = Path(__file__).parent
  digest = hashlib.sha256()
  extra = {"scaffold_event_buffer.py", "symbolic_sca2peg.py", "galil_clock.py", "read_blocks.py"}
  for path in sorted(root.glob("*.py")):
    if path.name.startswith(("scaffold_circuit", "fpp_", "dp_")) or path.name in extra:
      digest.update(path.name.encode() + b"\0" + path.read_bytes() + b"\0")
  return digest.hexdigest()


def generate(output, quantum, report, shared_cells=False, wrapper_cache=None,
             wrapper_only=False, rebuild_wrapper=False, lazy_round=False):
  timing = derive(quantum)
  report.update(timing=vars(timing), service=timing.service,
                slots=timing.service + 1, shared_cells=shared_cells, lazy_round=lazy_round)
  fingerprint = dict(signature=source_signature(), quantum=quantum, shared_cells=shared_cells)
  if wrapper_cache and wrapper_cache.exists() and not rebuild_wrapper:
    report["stage"] = "load finite FIFO artifact"
    print(json.dumps(report), flush=True)
    wrapper, metadata = scaffold_artifact.read(wrapper_cache)
    if metadata != fingerprint:
      raise ValueError("FIFO artifact differs from this source; use --rebuild-wrapper or another path")
  else:
    report["stage"] = "online source"
    print(json.dumps(report), flush=True)
    with share_expressions():
      circuit, source = build_online(quantum, shared_cells=shared_cells)
    ports = (circuit._label("online.ready", 0), circuit._label("online.event", 0), source.accepting)
    report["source"] = dict(labels=len(source.labels), pointers=len(source.pointers))
    report["stage"] = "finite FIFO wrapper"
    print(json.dumps(report), flush=True)
    del circuit
    gc.collect()
    with share_expressions():
      buffer_circuit, wrapper = buffer_source(source, *ports)
    del buffer_circuit, source
    gc.collect()
    if wrapper_cache:
      report["stage"] = "save finite FIFO artifact"
      print(json.dumps(report), flush=True)
      report["artifact_nodes"] = scaffold_artifact.write(wrapper, wrapper_cache, fingerprint)
  report["wrapper"] = dict(labels=len(wrapper.labels), pointers=len(wrapper.pointers))
  if wrapper_only:
    report.update(stage="finite FIFO artifact saved", artifact=str(wrapper_cache), emitted=False)
    return
  report["stage"] = "whole input round"
  print(json.dumps(report), flush=True)
  with share_expressions():
    packed = pack_service(wrapper, timing.service, lazy=lazy_round)
  del wrapper
  gc.collect()
  report["packed"] = dict(labels=len(packed.labels), pointers=len(packed.pointers))
  report["stage"] = "ordinary PEG emission"
  print(json.dumps(report), flush=True)
  temporary = output.with_suffix(output.suffix + ".partial")
  with temporary.open("w") as stream:
    for count, rule in enumerate(packed.iter_rules(), 1):
      stream.write(rule + "\n")
  temporary.replace(output)
  report.update(stage="emitted", rules=count, bytes=output.stat().st_size,
                output=str(output), emitted=True, raw_input_matches="not run")


def main():
  parser = argparse.ArgumentParser(description=__doc__)
  parser.add_argument("output", type=Path)
  parser.add_argument("--quantum", type=int, default=64)
  parser.add_argument("--memory-mib", type=int, default=16384)
  parser.add_argument("--report", type=Path)
  parser.add_argument("--shared-instruction-cells", action="store_true",
                      help="share mutually exclusive tape-move cells in finite programs")
  parser.add_argument("--wrapper-cache", type=Path,
                      help="save/load the exact finite FIFO DAG, with source fingerprint checking")
  parser.add_argument("--wrapper-only", action="store_true",
                      help="prepare the finite FIFO artifact without starting PEG generation")
  parser.add_argument("--rebuild-wrapper", action="store_true")
  parser.add_argument("--lazy-round", action="store_true")
  args = parser.parse_args()
  if args.memory_mib < 1:
    parser.error("positive compilation memory limit required")
  if args.wrapper_only and not args.wrapper_cache:
    parser.error("--wrapper-only requires --wrapper-cache")
  derive(args.quantum)
  resource.setrlimit(resource.RLIMIT_AS, (args.memory_mib * 1024 * 1024,) * 2)
  report, started = {}, time.monotonic()
  def interrupted(signum, frame):
    raise KeyboardInterrupt
  signal.signal(signal.SIGTERM, interrupted)
  try:
    generate(args.output, args.quantum, report, args.shared_instruction_cells,
             args.wrapper_cache, args.wrapper_only, args.rebuild_wrapper, args.lazy_round)
  except MemoryError:
    report.update(status="compilation memory limit exceeded", emitted=False)
    raise SystemExit(1)
  except KeyboardInterrupt:
    report.update(status="compilation interrupted", emitted=False)
    raise SystemExit(130)
  finally:
    report["seconds"] = time.monotonic() - started
    report["peak_rss_kib"] = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
    if args.report:
      args.report.write_text(json.dumps(report, indent=2) + "\n")
    print(json.dumps(report), flush=True)


if __name__ == "__main__":
  main()

File: scala/pal/src/main/scala/pal/GenerateOnlinePeg.scala
package pal

import java.nio.charset.StandardCharsets.UTF_8
import java.nio.file.{Files, Path, StandardCopyOption}
import java.security.MessageDigest
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicReference
import scala.collection.immutable.VectorMap
import scala.collection.mutable
import scala.jdk.CollectionConverters.*

/** Generate the FIFO/whole-round PAL candidate on original input.
  * No recognition budget, input expansion or word-length cap is accepted.
  * The output is replaced only after all ordinary PEG rules have been emitted.
  * Correctness remains subject to TRANSLATION_STRATEGY.md's source contracts.
  *
  * Port of `generate_online_peg.py`. The `.sca` representation remains
  * interoperable through [[ScaffoldArtifact]], but automatic generator-cache
  * reuse is language-specific because its signature covers Scala sources.
  */
object GenerateOnlinePeg {
  type Report = mutable.LinkedHashMap[String, Any]
  final case class Source(machine: Scaffold, ready: String, event: String)

  /** Construction hooks permit a tiny independent source in differential tests. */
  final case class Backend(
      signature: () => String = () => sourceSignature(),
      build: (Int, Boolean) => Source = (quantum, shared) => {
        val (circuit, machine) = ScaffoldCircuitGalil.buildOnline(quantum, sharedCells = shared)
        Source(machine, circuit.labelIds(("online.ready", 0)), circuit.labelIds(("online.event", 0)))
      },
      pack: (Scaffold, Int, Boolean) => Scaffold = (wrapper, service, lazyRound) =>
        ScaffoldEventBuffer.packService(wrapper, service, lazyRound)
  )

  def sourceSignature(sourceRoot: Path = scalaSourceDirectory()): String = {
    val digest = MessageDigest.getInstance("SHA-256")
    digest.update("lean4-peg/generate-online-peg/scala/v1\u0000".getBytes(UTF_8))
    val stream = Files.list(sourceRoot)
    val paths = try { stream.iterator().asScala.filter(_.getFileName.toString.endsWith(".scala")).toVector
      .sortBy(_.getFileName.toString) } finally { stream.close() }
    for (path <- paths) {
      digest.update(path.getFileName.toString.getBytes(UTF_8))
      digest.update(0.toByte)
      digest.update(Files.readAllBytes(path))
      digest.update(0.toByte)
    }
    digest.digest().map(byte => f"${byte & 0xff}%02x").mkString
  }

  private def scalaSourceDirectory(): Path = {
    val pythonRoot = GenerateScaffoldExamples.locateDirectory(Array.empty)
    val sourceRoot = pythonRoot.resolve("../../scala/pal/src/main/scala/pal").normalize()
    if (!Files.isDirectory(sourceRoot)) {
      throw new IllegalStateException(s"Scala source directory unavailable for cache provenance: $sourceRoot")
    }
    sourceRoot
  }

  private def timingFields(timing: GalilClock.Timing): VectorMap[String, Any] = {
    VectorMap("quantum" -> timing.quantum, "stage_factor" -> timing.stageFactor,
      "match_delay" -> timing.matchDelay, "move_slope" -> timing.moveSlope,
      "interval_overhead" -> timing.intervalOverhead, "predictability" -> timing.predictability)
  }

  /** Python's compact json.dumps formatting, preserving insertion order. */
  def reportJson(value: Any): String = value match {
    case entries: collection.Map[?, ?] => entries.iterator.map { case (key, item) =>
      PyFormat.jsonString(key.toString, ensureAscii = true) + ": " + reportJson(item)
    }.mkString("{", ", ", "}")
    case values: Seq[?] => values.map(reportJson).mkString("[", ", ", "]")
    case text: String => PyFormat.jsonString(text, ensureAscii = true)
    case value: Boolean => if (value) { "true" } else { "false" }
    case value: Int => value.toString
    case value: Long => value.toString
    case value: Double => PyFormat.floatRepr(value)
    case None => "null"
    case other => throw new IllegalArgumentException(s"unsupported report value: $other")
  }

  def reportPretty(value: Any, depth: Int = 0): String = {
    def collection(open: String, close: String, entries: Vector[String]): String = {
      if (entries.isEmpty) { open + close } else {
        open + "\n" + entries.map("  " * (depth + 1) + _).mkString(",\n") + "\n" + "  " * depth + close
      }
    }
    value match {
      case entries: scala.collection.Map[?, ?] =>
        collection("{", "}", entries.iterator.map { case (key, item) =>
          PyFormat.jsonString(key.toString, ensureAscii = true) + ": " + reportPretty(item, depth + 1)
        }.toVector)
      case entries: Seq[?] => collection("[", "]", entries.map(reportPretty(_, depth + 1)).toVector)
      case scalar => reportJson(scalar)
    }
  }

  def generate(output: Path, quantum: Int, report: Report, sharedCells: Boolean = false,
               wrapperCache: Option[Path] = None, wrapperOnly: Boolean = false,
               rebuildWrapper: Boolean = false, lazyRound: Boolean = false,
               backend: Backend = Backend(), progress: String => Unit = println): Unit = {
    val timing = GalilClock.derive(quantum)
    report("timing") = timingFields(timing)
    report("service") = timing.service
    report("slots") = timing.service + 1
    report("shared_cells") = sharedCells
    report("lazy_round") = lazyRound
    val fingerprint = VectorMap[String, Any]("signature" -> backend.signature(), "quantum" -> quantum, "shared_cells" -> sharedCells)
    def stage(name: String): Unit = { report("stage") = name; progress(reportJson(report)) }
    val wrapper = if (wrapperCache.exists(Files.exists(_)) && !rebuildWrapper) {
      stage("load finite FIFO artifact")
      val (cached, metadata) = ScaffoldArtifact.read(wrapperCache.get)
      if (metadata != fingerprint) {
        throw new IllegalArgumentException("FIFO artifact differs from this source; use --rebuild-wrapper or another path")
      }
      cached
    } else {
      stage("online source")
      val source = Expr.share { backend.build(quantum, sharedCells) }
      report("source") = VectorMap("labels" -> source.machine.labels.size, "pointers" -> source.machine.pointers.size)
      stage("finite FIFO wrapper")
      val (_, buffered) = Expr.share {
        ScaffoldEventBuffer.bufferSource(source.machine, source.ready, source.event, source.machine.accepting)
      }
      wrapperCache.foreach { path =>
        stage("save finite FIFO artifact")
        report("artifact_nodes") = ScaffoldArtifact.write(buffered, path, fingerprint)
      }
      buffered
    }
    report("wrapper") = VectorMap("labels" -> wrapper.labels.size, "pointers" -> wrapper.pointers.size)
    if (wrapperOnly) {
      report("stage") = "finite FIFO artifact saved"
      report("artifact") = wrapperCache.fold("None")(_.toString)
      report("emitted") = false
    } else {
      stage("whole input round")
      val packed = Expr.share { backend.pack(wrapper, timing.service, lazyRound) }
      report("packed") = VectorMap("labels" -> packed.labels.size, "pointers" -> packed.pointers.size)
      stage("ordinary PEG emission")
      val temporary = output.resolveSibling(output.getFileName.toString + ".partial")
      val stream = Files.newBufferedWriter(temporary, UTF_8)
      var count = 0L
      try {
        for (rule <- packed.iterRules()) { stream.write(rule); stream.write("\n"); count += 1 }
      } finally { stream.close() }
      Files.move(temporary, output, StandardCopyOption.REPLACE_EXISTING)
      report("stage") = "emitted"
      report("rules") = count
      report("bytes") = Files.size(output)
      report("output") = output.toString
      report("emitted") = true
      report("raw_input_matches") = "not run"
    }
  }

  final case class Options(output: Path, quantum: Int = 64, memoryMib: Int = 16384,
                           report: Option[Path] = None, sharedCells: Boolean = false,
                           wrapperCache: Option[Path] = None, wrapperOnly: Boolean = false,
                           rebuildWrapper: Boolean = false, lazyRound: Boolean = false)

  def parseArgs(args: Array[String]): Options = {
    var output: Option[Path] = None
    var quantum = 64
    var memoryMib = 16384
    var report: Option[Path] = None
    var shared = false
    var cache: Option[Path] = None
    var only = false
    var rebuild = false
    var lazyRound = false
    var index = 0
    def argument(flag: String): String = {
      index += 1
      if (index >= args.length) { throw new IllegalArgumentException(s"$flag requires a value") }
      args(index)
    }
    while (index < args.length) {
      args(index) match {
        case "--quantum" => quantum = argument("--quantum").toInt
        case "--memory-mib" => memoryMib = argument("--memory-mib").toInt
        case "--report" => report = Some(Path.of(argument("--report")))
        case "--shared-instruction-cells" => shared = true
        case "--wrapper-cache" => cache = Some(Path.of(argument("--wrapper-cache")))
        case "--wrapper-only" => only = true
        case "--rebuild-wrapper" => rebuild = true
        case "--lazy-round" => lazyRound = true
        case value if !value.startsWith("-") && output.isEmpty => output = Some(Path.of(value))
        case value => throw new IllegalArgumentException(s"unrecognized argument: $value")
      }
      index += 1
    }
    if (memoryMib < 1) { throw new IllegalArgumentException("positive compilation memory limit required") }
    if (only && cache.isEmpty) { throw new IllegalArgumentException("--wrapper-only requires --wrapper-cache") }
    GalilClock.derive(quantum)
    Options(output.getOrElse(throw new IllegalArgumentException("output path required")), quantum, memoryMib, report,
      shared, cache, only, rebuild, lazyRound)
  }

  /** The JVM reserves its heap at startup: enforce --memory-mib by requiring
    * a JVM -Xmx at or below it. This caps heap, rather than Python's RLIMIT_AS.
    */
  def main(args: Array[String]): Unit = {
    if (args.contains("--help") || args.contains("-h")) {
      println("Usage: GenerateOnlinePeg OUTPUT [--quantum N] [--memory-mib N] [--report PATH]\n" +
        "  [--shared-instruction-cells] [--wrapper-cache PATH] [--wrapper-only]\n" +
        "  [--rebuild-wrapper] [--lazy-round]\n" +
        "Generate ordinary PEG rules on original input. --memory-mib requires a JVM -Xmx at or below the given MiB; it limits heap, not virtual address space.")
      return
    }
    val options = parseArgs(args)
    if (Runtime.getRuntime.maxMemory() > options.memoryMib.toLong * 1024 * 1024) {
      throw new IllegalArgumentException(s"start the JVM with -Xmx${options.memoryMib}m or less to enforce --memory-mib")
    }
    val report = mutable.LinkedHashMap.empty[String, Any]
    val started = System.nanoTime()
    val finalized = new AtomicBoolean(false)
    val published = new AtomicReference[VectorMap[String, Any]](VectorMap.empty)
    val reportLock = new Object()
    def finishReport(interrupted: Boolean): Unit = reportLock.synchronized {
      if (finalized.compareAndSet(false, true)) {
        var snapshot = if (interrupted) { published.get() } else { VectorMap.from(report) }
        if (interrupted) { snapshot = snapshot.updated("status", "compilation interrupted").updated("emitted", false) }
        snapshot = snapshot.updated("seconds", (System.nanoTime() - started) / 1e9)
        val status = Path.of("/proc/self/status")
        val peak = if (Files.isReadable(status)) {
          Files.readAllLines(status).asScala.find(_.startsWith("VmHWM:")).map(_.trim.split("\\s+")(1).toLong).getOrElse(0L)
        } else { None }
        snapshot = snapshot.updated("peak_rss_kib", peak)
        options.report.foreach(path => Files.writeString(path, reportPretty(snapshot) + "\n", UTF_8))
        println(reportJson(snapshot))
      }
    }
    // A normal JVM SIGTERM keeps its platform exit status (typically 143), but
    // this hook still publishes an interruption report from an immutable stage
    // snapshot. No internal signal API or concurrent mutable-map iteration.
    val shutdownHook = new Thread(() => finishReport(interrupted = true), "generate-online-peg-report")
    Runtime.getRuntime.addShutdownHook(shutdownHook)
    var exitCode = 0
    try {
      generate(options.output, options.quantum, report, options.sharedCells, options.wrapperCache,
        options.wrapperOnly, options.rebuildWrapper, options.lazyRound, progress = line => {
          published.set(VectorMap.from(report))
          println(line)
        })
    } catch {
      case _: OutOfMemoryError =>
        report("status") = "compilation memory limit exceeded"
        report("emitted") = false
        exitCode = 1
      case _: InterruptedException =>
        report("status") = "compilation interrupted"
        report("emitted") = false
        Thread.currentThread().interrupt()
        exitCode = 130
    } finally {
      finishReport(interrupted = false)
      try { Runtime.getRuntime.removeShutdownHook(shutdownHook) }
      catch { case _: IllegalStateException => () }
    }
    if (exitCode != 0) { sys.exit(exitCode) }
  }
}
